[WIP] Implement dynamically callable types (`@dynamicCallable`).

dan-zheng opened the pull request:
This proposal introduces the `@dynamicCallable` attribute, which enables
nominal types to be "callable" via a simple syntactic sugar.

This is the implementation of the Swift evolution proposal here:
https://github.com/apple/swift-evolution/pull/858.
Read there for more details.

---

`@dynamicCallable` is a follow-up to [SE-0195 "Dynamic Member Lookup"](https://github.com/apple/swift-evolution/blob/master/proposals/0195-dynamic-member-lookup.md).

Forum discussion regarding `@dynamicCallable`:
  - [Pitch #1: Introduce user-defined dynamically "callable" types](https://forums.swift.org/t/pitch-introduce-user-defined-dynamically-callable-types/7038)
  - [Pitch #2: Introduce user-defined dynamically “callable” types](https://forums.swift.org/t/pitch-2-introduce-user-defined-dynamically-callable-types/7112)
  - [Pitch #3: Introduce user-defined dynamically “callable” types](https://forums.swift.org/t/pitch-3-introduce-user-defined-dynamically-callable-types/12232)

Proposal review and acceptance rationale:
  - [Proposal review: SE-216: User-defined dynamically callable types](https://forums.swift.org/t/se-0216-user-defined-dynamically-callable-types/13615)
  - [[Accepted] SE-216: User-defined dynamically callable types](https://forums.swift.org/t/accepted-se-216-user-defined-dynamically-callable-types/14110)

dan-zheng committed:
Implement dynamically callable types (`@dynamicCallable`).

This proposal introduces the `@dynamicCallable` attribute, which enables
nominal types to be "callable" via a simple syntactic sugar.

This is the implementation of the Swift evolution proposal here:
https://github.com/apple/swift-evolution/pull/858.
Read there for more details.

dan-zheng committed:
Remove SWIFT_ENABLE_TENSORFLOW comments.

dan-zheng commented:
Todo:
* Add a `-emit-silgen` test to verify that dynamic calls resolve to the correct method (either the `withArguments:` or `withKeywordArguments:` method).
  * Currently, there's only a name binding test ([test/NameBinding/dynamic-callable.swift](https://github.com/dan-zheng/swift/blob/adb7648ded947183ebc003d7e378db69e10f82cf/test/NameBinding/dynamic-callable.swift)), which isn't sufficient for testing this.

dan-zheng committed:
Add @dynamicCallable SIL gen test.

This test verifies that dynamic calls resolve to the correct method.

dan-zheng committed:
Add comments to @dynamicCallable SIL gen test.

dan-zheng committed:
[@dynamicCallable] Minor clean up.

- Use known identifiers `Id_dynamicallyCall`, etc. in CSSimplify.
- Update doc comment in CSDiag.

dan-zheng commented:
@swift-ci Please test

dan-zheng commented:
@swift-ci Please test

dan-zheng commented:
@swift-ci Please smoke test

tkremenek commented:
@swift-ci build toolchain

dan-zheng committed:
[NFC] Edit formatting/comments for @dynamicCallable.

dan-zheng commented on lib/Sema/CSApply.cpp:
```diff
@@ -7601,67 +7601,147 @@ Expr *ExprRewriter::finishApply(ApplyExpr *apply, Type openedType,
-      auto instanceType = metaType->getInstanceType();
-      isExistentialType = instanceType->isExistentialType();
-    }
+  if (auto metaTy = cs.getType(fn)->getAs<AnyMetatypeType>()) {
```
Note: the diff at this part of the file isn't actually that big.
I simply moved the metatype-handling code (lines 7604-7655) into an if-block.

slavapestov replied:
This method is already quite long. Can you split off this part into a separate method?

dan-zheng commented:
@tkremenek I think I nullified your "build toolchain" command by pushing a new commit. Sorry about that.
@swift-ci Please smoke test

dan-zheng commented:
@swift-ci Please smoke test

dan-zheng commented:
@swift-ci Please test

tkremenek commented:
@swift-ci build toolchain

tkremenek commented:
Neither of the download links for the toolchains appear to work.

dan-zheng commented:
@tkremenek 
I'm able to download the toolchains by navigating to the build job pages and downloading the artifact there.
- macOS: [build job page](https://ci.swift.org/job/swift-PR-toolchain-osx/33/), [toolchain download link](https://ci.swift.org/job/swift-PR-toolchain-osx/33/artifact/branch-master/swift-PR-16980-33-osx.tar.gz)
- Linux: [build job page](https://ci.swift.org/job/swift-PR-toolchain-Linux/26/), [toolchain download link](https://ci.swift.org/job/swift-PR-toolchain-Linux/26/artifact/branch-master/swift-PR-16980-26-ubuntu16.04.tar.gz)

shahmishal commented:
@swift-ci comment updated with correct download links

dan-zheng committed:
Remove "non-generic" requirement from @dynamicCallable diagnostic.

At some point, the @dynamicCallable implementation did not support
generic `dynamicallyCall` methods.
However, generic methods are now supported.

dan-zheng commented:
@swift-ci Build toolchain

shahmishal commented:
@swift-ci Build toolchain

tkremenek commented:
@swift-ci build toolchain

shahmishal commented:
@swift-ci build toolchain

slavapestov commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4254,6 +4253,133 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
+  auto calculate = [&]() -> DynamicCallableMethods {
+    // If this is a protocol composition, check if any of the protocols have the
+    // attribute.
+    if (auto protocolComp = canType->getAs<ProtocolCompositionType>()) {
```
Use `dyn_cast<T>(canType)` instead of `canType->getAs<T>`

slavapestov commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4254,6 +4253,133 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
+    // attribute.
+    if (auto protocolComp = canType->getAs<ProtocolCompositionType>()) {
+      DynamicCallableMethods methods;
+      for (auto protocolType : protocolComp->getMembers()) {
```
Protocol compositions can also have class members -- did you test that case?

slavapestov commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4254,6 +4253,133 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
+    // look up the methods.
+    for (auto p : nominal->getAllProtocols())
+      if (p->getAttrs().hasAttribute<DynamicCallableAttr>())
+        return lookupDynamicCallableMethods(type, CS, locator, error);
```
What about protocols that refine other protocols?

dan-zheng replied:
There's a protocol refinement test [here](https://github.com/dan-zheng/swift/blob/fe91f19de1c273e6179aaa08a71fa145ae314d41/test/NameBinding/dynamic-callable.swift#L197).
I'd be happy to add more tests, if you have suggestions.

slavapestov commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4372,6 +4498,114 @@ ConstraintSystem::simplifyApplicableFnConstraint(
     return simplified;
   }
 
+  // Handle @dynamicCallable applications.
```
This is a lot of new code for a single method -- maybe factor it out into a new method?

slavapestov commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4372,6 +4498,114 @@ ConstraintSystem::simplifyApplicableFnConstraint(
+      Type dictValueType =
+        ProtocolConformanceRef::getTypeWitnessByName(
+          argType, *conformance, ctx.Id_Value, &TC)
+        ->getDesugaredType();
```
Why are you desugaring it?

slavapestov commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4372,6 +4498,114 @@ ConstraintSystem::simplifyApplicableFnConstraint(
+      Type arrayElementType =
+        ProtocolConformanceRef::getTypeWitnessByName(
+          argType, *conformance, ctx.Id_ArrayLiteralElement, &TC)
+        ->getDesugaredType();
```
Why are you desugaring it?

slavapestov commented on test/NameBinding/dynamic-callable.swift:
```diff
@@ -0,0 +1,373 @@
+    return arguments[0].1
+  }
+}
+func testGenericType5<T>(a: CallableGeneric5<T>) -> Double {
```
Please also test dynamic calls on an archetype:

```
func f<T>(_ t: T) where T : ... {
  t.foo()
}
```

With protocol and class constraints on `T`.

dan-zheng replied:
Sorry for the late response.

---
I added [such a test](https://github.com/apple/swift/pull/16980/commits/fe91f19de1c273e6179aaa08a71fa145ae314d41#diff-e215fa74b04a179ecc272f8d774d7199R382) and it failed - thank you for the catch!

I'll take a closer look this weekend. It seems `@dynamicMemberLookup` has the same issue with archetypes:
```swift
@dynamicMemberLookup
class MemberLookup {
  subscript(dynamicMember name: String) -> Int { return 1 }
}

func testPass(a: MemberLookup) -> Int {
  return a.foo // ok
}
func testFail<T : MemberLookup>(a: T) -> Int {
  return a.foo // error: value of type 'T' has no member 'foo'
}
```

slavapestov commented on test/NameBinding/dynamic-callable.swift:
```diff
@@ -0,0 +1,373 @@
+func testExistential2(a: AnyObject & CallableProtocol,
+                      b: SubProtocol,
+                      c: ProtocolComp & AnyObject,
+                      d: CallableProtocol & KeywordCallableProtocol) {
```
Also please test `SomeProto & SomeClass`, or `SomeProto & SomeTypeAlias` with `typealias T = ...`

dan-zheng committed:
Merge remote-tracking branch 'origin/master' into dynamic-callable

dan-zheng committed:
Minor cleanup.

- Refactor CSSimplify @dynamicCallable logic into a helper function.

Address comments from @slavapestov.

dan-zheng committed:
Add @dynamicCallable tests.

- Add tests for protocol compositions with class members and archetypes, as
  suggested by @slavapestov.
- TODO: fix failing archetype test.

dan-zheng committed:
Add @dynamicCallable tests: protocol constrained/refined archetypes.

- Add `testConstrainedClassType` and `testRefinedProtocolType`.

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci build toolchain

slavapestov commented:
@dan-zheng does @dynamicMemberLookup support the same thing where the attribute could be on an adopted protocol or superclass? If so, can we refactor the implementations to share code?

slavapestov commented:
@dan-zheng Great to hear its on your list, thanks!

dan-zheng commented:
Update: I plan to look into the current failures this weekend.
The main issue is dynamic calls on archetypes don't work. Another issue is that `dynamicallyCall` methods currently must be concrete, they don't work as protocol requirements.
`@dynamicMemberLookup` also has both of these issues.

slavapestov commented on lib/Sema/CSApply.cpp:
```diff
@@ -7623,67 +7623,147 @@ Expr *ExprRewriter::finishApply(ApplyExpr *apply, Type openedType,
+  // At this point, all other AppyExpr cases have been handled.
+  auto isDynamicCallable = [&](Expr *base) {
+    auto type = cs.getType(base);
+    return type && cs.DynamicCallableCache.count(type->getCanonicalType());
```
When is type ever null?

dan-zheng replied:
Thanks for the catch, `type` should never be null.
Removed `type &&` from condition in https://github.com/apple/swift/pull/16980/commits/1329ba1ad003d1706a66dd298d2079ea24f57d81.

slavapestov commented on lib/Sema/CSApply.cpp:
```diff
@@ -7623,67 +7623,147 @@ Expr *ExprRewriter::finishApply(ApplyExpr *apply, Type openedType,
+    auto type = cs.getType(base);
+    return type && cs.DynamicCallableCache.count(type->getCanonicalType());
+  };
+  assert(isDynamicCallable(fn) && "Expected a valid @dynamicCallable type");
```
It might make sense to split off the dynamic callable handling into a separate method also, making finishApply() nice and short

dan-zheng replied:
Done in https://github.com/apple/swift/pull/16980/commits/1329ba1ad003d1706a66dd298d2079ea24f57d81.

slavapestov commented on test/NameBinding/dynamic-callable-silgen.swift:
```diff
@@ -0,0 +1,25 @@
+// RUN: %target-swift-frontend -emit-silgen -verify %s | %FileCheck %s
```
SILGen tests should be located in test/SILGen/

dan-zheng replied:
Done in https://github.com/apple/swift/pull/16980/commits/1329ba1ad003d1706a66dd298d2079ea24f57d81.

slavapestov commented on test/NameBinding/dynamic-callable.swift:
```diff
@@ -0,0 +1,396 @@
+// RUN: %target-swift-frontend -typecheck -verify %s
```
I would instead put this in test/attr/

dan-zheng replied:
Done in https://github.com/apple/swift/pull/16980/commits/1329ba1ad003d1706a66dd298d2079ea24f57d81.

I assume `test/NameBinding/dynamic-member-lookup.swift` should also be moved to `test/attr`. I'll remember to do so in a later PR (when I unite code between `@dynamicCallable` and `@dynamicMemberLookup`).

dan-zheng committed:
Minor edits.

Address feedback from @slavapestov.
Move @dynamicCallable tests to more sensible locations.

dan-zheng committed:
Merge branch 'master' into dynamic-callable

dan-zheng committed:
Support @dynamicCallable archetypes.

Support @dynamicCallable method lookup for archetypes.
Generalize code for protocol composition types.

dan-zheng commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4302,33 +4302,51 @@ getDynamicCallableMethods(Type type, ConstraintSystem &CS,
+  auto calculate = [&]() -> DynamicCallableMethods {
+    // If this is an archetype type, check if any types it conforms to
+    // (superclass or protocols) have the attribute.
+    if (auto archetype = dyn_cast<ArchetypeType>(canType)) {
```
@slavapestov Could you verify whether the code here for handling archetypes is sufficient?

To get `@dynamicCallable` methods from an archetype, I look for the methods in the archetype's `getConformsTo()` protocols and superclass (if defined). Are there cases that aren't handled by this?

slavapestov replied:
This looks reasonable.

At this point, the 'component types' calculation and superclass walk duplicates work that's being done in name lookup. It would be great if at some later point in time we could come up with a better abstraction here!

dan-zheng replied:
Thanks for the astute observation!
The "type walking" logic in `getDynamicCallableMethods` is gnarly and should definitely be fixed + simplified if possible. I'm not sure how to best do this at the moment.

dan-zheng commented:
@swift-ci Please smoke test

dan-zheng committed:
Rename/edit @dynamicCallable SILGen test.

Following underscore filename naming convention.

dan-zheng commented:
@swift-ci Please smoke test

dan-zheng commented:
@swift-ci Please smoke test

dan-zheng commented:
@slavapestov @rudkx
FYI to reviewers: this PR is ready for review now.

Previous issues that have been fixed:
* Dynamic calls now work on archetypes (class types, protocol types, protocol-constrained class types, etc)
* The `dynamicallyCall` methods now work as protocol requirements all the time. Previously, they didn't work as protocol requirements when defined on archetypes.

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4239,6 +4238,267 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
+  // keyword arguments method.
+  bool useKwargsMethod = methods.argumentsMethod == nullptr;
+  if (!useKwargsMethod) {
+    for (auto param : func1->getParams()) {
```
Nit: you can simplify this loop into: 

```c++
llvm::find_if(func1->getParams(), [](AnyFunctionType::Param param) { return param.hasLabel(); });
```

dan-zheng replied:
Thanks! Done in https://github.com/apple/swift/pull/16980/commits/a982340e8d3494dc6c7924b63c2f6f2da1353142.

dan-zheng replied:
Actually, this change caused `attr/attr_dynamic_callable.swift` to fail for some reason.
`llvm::find_if` produces different behavior than the original for-loop code.

@xedin, do you have any idea what's going on?

dan-zheng committed:
[@dynamicCallable] Minor edit.

Address comment by @xedin.
Refactor for-loop into llvm::find_if.

dan-zheng committed:
[@dynamicCallable] Minor edit.

Use for-loop instead of `llvm::find_if`, undoing the previous commit:
https://github.com/apple/swift/pull/16980/commits/a982340e8d3494dc6c7924b63c2f6f2da1353142

The correct usage of `llvm::find_if` is somewhat convoluted and less readable imo:
```
useKwargsMethod = llvm::find_if(func1->getParams(),
    [](AnyFunctionType::Param param) {
        return param.hasLabel();
    }) != func1->getParams().end();
```

dan-zheng commented:
I undid https://github.com/apple/swift/pull/16980/commits/a982340e8d3494dc6c7924b63c2f6f2da1353142, which caused errors explained [here](https://github.com/apple/swift/pull/16980#discussion_r203575092).
@swift-ci Please smoke test

dan-zheng committed:
Merge branch 'master' into dynamic-callable

rudkx commented:
@swift-ci Please test

rudkx commented:
@swift-ci Please test source compatibility

slavapestov commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4252,6 +4251,267 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
+
+      // If this type has the attribute on it, then look up the methods.
+      if (nominal->getAttrs().hasAttribute<DynamicCallableAttr>())
+        return lookupDynamicCallableMethods(type, CS, locator, error);
```
One minor nitpick. It's not apparent to me from looking at this code, but getAllProtocols() on a class includes protocols conformed to by superclasses, so I'm wondering if you're redundantly walking protocols of superclasses here when you don't need to.

dan-zheng replied:
Hmm. I don't believe this is the case because `getAllProtocols()` is checked before the loop, meaning all protocols conformed to by superclasses are checked once. If none of those protocols have the `@dynamicCallable` attribute, the loop will recursively check superclasses of the class without rechecking protocols.

dan-zheng commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4252,6 +4251,267 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
+    // look up the methods.
+    for (auto p : nominal->getAllProtocols())
+      if (p->getAttrs().hasAttribute<DynamicCallableAttr>())
+        return lookupDynamicCallableMethods(type, CS, locator, error);
```
Currently, `for (auto p : nominal->getAllProtocols())` iterates over all protocols and looks up methods in the first protocol that has the `@dynamicCallable` attribute.

Instead, should `calculateForComponentTypes` be called on the all protocols with the `@dynamicCallable` attribute? That seems more correct.

rudkx commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4252,6 +4251,267 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
+    ConstraintSystem &CS, Type type1, Type type2,
+    ConstraintLocatorBuilder locator,
+    ConstraintSystem::TypeMatchOptions flags) {
+  using SolutionKind = ConstraintSystem::SolutionKind;
```
This really doesn't behave like normal overload resolution. For example we fail to type-check the following, which I believe a user might normally expect to resolve to the overload with keyword arguments:

```
@dynamicCallable
struct S {
  func dynamicallyCall(withArguments: [String]) -> Int {
    fatalError()
  }
  func dynamicallyCall(withKeywordArguments: [String : String]) -> Float {
    fatalError()
  }
}

let s = S()
let _: Float = s("hi")
```

dan-zheng replied:
This is a good point. I describe the problems with the lookup-based approach in more detail [here](https://github.com/apple/swift/pull/16980#issuecomment-410142870). Do you have ideas for a better method resolution approach?

rudkx commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4252,6 +4251,267 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
+lookupDynamicCallableMethod(Type type, ConstraintSystem &CS,
+                            const ConstraintLocatorBuilder &locator,
+                            Identifier argumentName, bool hasKeywordArgs,
+                            bool &error) {
```
It's not clear to me what value the separate error parameter adds here. Another option might be to have `getDynamicCallableMethods` return an llvm::Optional and use the existence or lack of a value as an indication of the success or failure, unless you want to carry more semantic information in the form of a richer error type than a bool.

dan-zheng replied:
Thanks for pointing this out. I'll use `llvm::Optional` instead, same for `hasValidDynamicCallableMethod` in `TypeCheckAttr.cpp` which currently also takes a `bool &error` argument.

rudkx reviewed (commented):
Here's some initial feedback. I'm still looking over some of the details.

lattner commented:
Hey @dan-zheng , are you still working on this?  I'd really love to see this land.

dan-zheng commented:
I sincerely apologize for the long absence.

As @slavapestov and @rudkx pointed out, the current `@dynamicCallable` implementation has some big problems (many of which are shared with `@dynamicMemberLookup`). Dynamic call resolution doesn’t behave like standard overload resolution: it’s ad-hoc, lookup based, and frustratingly doesn’t utilize the constraint system.

At a high level, I'd say `@dynamicCallable` method lookup should work just like protocol requirement lookup (please comment if you disagree). The current lookup based process cannot achieve this: overloaded `dynamicallyCall` methods are not at all supported, leading the following code to fail:
```swift
// error: @dynamicCallable attribute does not support ambiguous method 'dynamicallyCall(withArguments:)'
@dynamicCallable
protocol Callable {
  func dynamicallyCall(withArguments: [Int]) -> Int
}
extension Callable {
  func dynamicallyCall(withArguments: [Int]) -> Int {
    return 1
  }
}
```

This is unexpected behavior in my opinion. Rather than iterating on the current lookup based approach, I think a constraint-system based solution would be more robust. Consider the following example:
```swift
@dynamicCallable
protocol Callable {
  // Overloaded `@dynamicCallable` methods should be supported, imo.
  func dynamicallyCall(withArguments: [Int]) -> Int
  func dynamicallyCall(withArguments: [Float]) -> Float
}
extension Callable {
  func dynamicallyCall(withArguments: [Int]) -> Int {
    return 1
  }
}
func foo(x: Callable) -> Float {
  return x(1, 2, 3) // how can we resolve this call?
}
```

One source of friction is that dynamic calls are not an exact sugar for direct calls from the constraint system's perspective. (For the sake of discussion, I'll focus on the `withArguments` method and ignore `withKeywordArguments` for now.)

In direct calls (e.g. `x.dynamicallyCall(withArguments: [1, 2, 3])`), only one type variable is created for the `ArrayExpr` element type.
In dynamic calls (e.g. `x(1, 2, 3)`), one type variable is created for each element of the argument tuple (three total here).

In `CSSimplify.cpp`, the current logic is:
* Find a suitable `dynamicallyCall(withArguments:)` method via lookup. Overloaded methods are not supported. Let `$A` and `$R` respectively be the argument and result type of the method. `$A` is an `ArraySliceType` with base type `$X`.
* For each type variable in the argument expression tuple (e.g. `$T0`, `$T1`, ...), add an `ArgumentConversion` constraint from it to `$X`. Add a `Bind` constraint from the result type variable to `$R`.

As mentioned above, this approach is problematic. Here's one other idea I had:
* Look up all `dynamicallyCall(withArguments:)` methods. For each of them, apply constraints like the current logic, then add a disjunction constraint of all those sub-constraints. I don't know if it's possible to "pair up" the argument/result type constraints per method, though, so that either both or neither constraints solve.
* Extend the current lookup based approach to support overloads. I'm not sure how to rank overloads though, after all methods have been found.

Constraint system experts: do you have any ideas/suggestions for fixing the current implementation? I'm on standby because I don't know how to best resolve overloads, mimicking protocol requirement resolution. I'm happy to test suggestions and to make a final push to complete the implementation.

dan-zheng commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -4252,6 +4251,267 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
+        if (CS.DynamicCallableCache.count(protocol->getCanonicalType()))
+          decl = protocol->getAnyNominal();
+    }
+    CS.TC.diagnose(decl->getLoc(), diag::missing_dynamic_callable_kwargs_method,
```
Thanks for pointing this out. I'll look into improving diagnostics.

dan-zheng commented:
Here's a more scoped question: how does overload resolution work in the example below?
```swift
protocol A {
  func foo(_ x: [Int]) -> Int
}
extension A {
  func foo(_ x: [Int]) -> Int { return 1 }
}
func test(a: A) -> Int {
  return a.foo([]) // how does this resolve to the protocol extension method?
}
```

rudkx commented:
In an example like the one you've been given, we type check the expression using both declarations of `foo`. Both result in valid solutions to type checking, so we disambiguate by comparing the solutions in `compareSolutions` of `CSRanking.cpp`.

That actually selects the protocol method, which dynamically dispatches to the appropriate method. So we don't statically resolve to the extension method in this case. If you had a type that extends `A` and provided an implementation of `foo`, and passed that to `test`, we would call that type's implementation of `foo` from `test`.

dan-zheng commented:
@rudkx Thanks for the explanation!

The `@dynamicCallable` method resolution logic could be changed to select protocol (and class?) methods when possible, to avoid the need to statically resolve methods.

I took a lot at `compareSolutions` in `CSRanking.cpp`; it probably doesn't make sense to call `compareSolutions` directly in `CSSimplify.cpp` since it's from a different part of the constraint system code path.

Improved `@dynamicCallable` method resolution logic should support overloaded methods and work even in trickier situations like the one described [here](https://github.com/apple/swift/pull/16980#discussion_r204202532). Here are some pseudo-steps to implement this:
1. In `CSSimplify.cpp`, check if the "called" nominal type has or somehow "inherits" the `@dynamicCallable` attribute. Then, look up and cache all `dynamicallyCall(withArguments:)` and `dynamicallyCall(withKeywordArguments:)`methods within the nominal type.
2. In `CSSimplify.cpp`, resolve the correct method, or emit an diagnostic (no compatible methods, ambiguous methods, etc). This requires somehow checking whether the argument/result type of the dynamic call expression can convert/bind the argument/result type of each method.
    * The resolution logic should prefer protocol (and class?) methods when possible to leverage dynamic dispatch.
    * The resolution logic should be aware that dynamic calls with no labels can still resolve to the `withKeywordArguments` method (e.g. `s("x")` in the code [here](https://github.com/apple/swift/pull/16980#discussion_r204202532)).
3. If resolution failed and a diagnostic has been emitted, we're done. If not, in `CSApply.cpp`, it must be possible to access the resolved method for expression rewriting.

The big question here is how to implement the resolution logic in step 2. I forget how exactly `CS.matchTypes` works, but perhaps it can be used in the following approach:
* Iterate through all candidate methods.
    * If type-checking a dynamic call with labels, candidates are only `withKeywordArguments:` methods.
    * If type-checking a dynamic call with no labels, candidates are both `withArguments:` and `withKeywordArguments:` methods.
* Filter the candidates, using `CS.matchTypes` to see if each method's argument/result types are compatible with the dynamic call's argument and result types.
    * If there are no valid candidates, emit a diagnostic.
    * If there is exactly one valid candidate, somehow mark it as resolved so that it's accessible in `CSApply.cpp`.
    * If there are multiple valid candidates, things are trickier.
        * If there are two methods with the same type, where one is a protocol requirement and the other is a concrete method (e.g. non protocol requirement), prefer the protocol requirement. (Does this rule always hold or need extra qualification?)
         * What should be done if there are multiple protocol requirements with the same type? (Tentative answer: randomly select one of them, if they have the exact same type. I believe this is the current logic for compositions of protocol that have identical required methods.)
         * What if there are multiple valid class methods where the classes inherit from one another? Does there need to be explicit logic to select the innermost subclass method, or is it possible to leverage dynamic dispatch?

Please reply if you have feedback on this approach, especially if you think some part won't work or if you have a better idea. I have some priorities now but I'll be free to experiment and work on this next Friday and later.

dan-zheng committed:
Merge branch 'master' into dynamic-callable

dan-zheng committed:
Merge branch 'master' into dynamic-callable

slavapestov commented:
@dan-zheng Comparing solutions happens after all constraints have been solved.

Also you should not emit diagnostics during constraint simplification because we can explore different parts of the solution space as we visit disjunctions, and discard invalid paths.

dan-zheng commented:
@slavapestov Thanks for sharing that info.
I edited my approach, replacing "emit a diagnostic during constraint simplification" with "return `SolutionKind::Error`".

dan-zheng committed:
[WIP] Reimplement @dynamicCallable resolution.

Reimplement dynamic call resolution using the approach described in:
https://github.com/apple/swift/pull/16980#issuecomment-410467807

This commit is a WIP, I'm pushing it to transfer code between machines.
I'll refactor things later.

dan-zheng commented:
@slavapestov Yes, I do plan to squash commits after the code is fixed.
Unfortunately, [my approach](https://github.com/apple/swift/pull/16980#issuecomment-410467807) for fixing constraint simplification didn't quite work, this part in particular:
> Filter the candidates, using `CS.matchTypes` to see if each method's argument/result types are compatible with the dynamic call's argument/result types.

 It's expected that all failing calls to `matchTypes` are immediately followed with `return SolutionKind::Error` (because `matchTypes` mutates the `ConstraintSystem`). This is problematic for the approach.

I'm not sure how to work around this: is it possible to clone/backtrack `ConstraintSystem` instances? Bigger changes to the approach may be necessary, advice would be much appreciated.

slavapestov commented:
@dan-zheng You can introduce disjunctions, which is the mechanism used to implement method overloads. @rudkx can give you more pointers.

dan-zheng commented:
@slavapestov @rudkx Do "union constraints" exist, in addition to disjunction constraints?

I need to emulate something like:
```
Call site type variables: (T, U) -> R. I need to work with these somehow.
candidate1: (A1, B1) -> R1
candidate2: (A2, B2) -> R2

Evaluate: ((T match A1) /\ (U match B1) /\ (R match R1)) \/ ((T match A2) /\ (U match B2) /\ (R match R2))
```

Advice on modeling this would be appreciated!

rudkx commented:
@dan-zheng That's what our disjunction constraints do when paired to an `ApplicableFunction` constraint. For example for:

```
func foo(_ i: Int) -> Int { return i }
func foo(_ d: Double) -> Double { return d }

func test() {
  _ = foo(1)
}
```

we generate the following constraints for the call to `foo(1)` from `generateConstraints()`:

```
  disjunction [[locator@0x10c806a00 [OverloadedDeclRef@/tmp/overload.swift:5:7]]]:$T1 bound to decl overload.(file).foo@/tmp/overload.swift:1:6 : (Int) -> Int at /tmp/overload.swift:1:6 [[locator@0x10c806a00 [OverloadedDeclRef@/tmp/overload.swift:5:7]]]; or $T1 bound to decl overload.(file).foo@/tmp/overload.swift:2:6 : (Double) -> Double at /tmp/overload.swift:2:6 [[locator@0x10c806a00 [OverloadedDeclRef@/tmp/overload.swift:5:7]]];
  $T2 literal conforms to ExpressibleByIntegerLiteral [[locator@0x10c806bd8 [IntegerLiteral@/tmp/overload.swift:5:11]]];
  ($T2) -> $T3 applicable fn $T1 [[locator@0x10c806d10 [Call@/tmp/overload.swift:5:7 -> apply function]]];
  $T3 conv $T0 [[locator@0x10c806de0 [Assign@/tmp/overload.swift:5:5]]];
```

The solver will then iterate over the options from the disjunction, binding each to `$T1` and `simplify()`ing  the `ApplicableFunction` constraint which produces the argument conversions from `$T2` to a newly create type variable for the argument to the option in the disjunction.

dan-zheng commented:
@rudkx Thanks!

I implemented `@dynamicCallable` method resolution using a disjunction paired with an `ApplicableFunction` constraint (unpolished code in [this commit](https://github.com/dan-zheng/swift/commit/b5c091a794df6ea3c5e79ae5d2131415da52557b#diff-69306f7b0822cf52e8c0639307ad7407R4591)) but ran into some road blocks.

Specifically, I generate an overload set disjunction in `simplifyDynamicCallableApplicableFunction`, following the example you shared above.

Code:
```swift
@dynamicCallable
struct Callable {
  func dynamicallyCall(withArguments: [Int]) {}
  func dynamicallyCall(withArguments: [Float]) {}
}
func test(x: Callable) {
  x(1, 2)
}
```

The constraints look like:
```
Type Variables:
  $T0 [lvalue allowed] as Callable @ locator@0x7ff540021000 [DeclRef@d2.swift:7:3]
  $T1 fully_bound literal=3 involves_type_vars bindings={(subtypes of) (default from ExpressibleByIntegerLiteral) Int} @ locator@0x7ff540021080 [IntegerLiteral@d2.swift:7:5]
  $T2 fully_bound literal=3 involves_type_vars bindings={(subtypes of) (default from ExpressibleByIntegerLiteral) Int} @ locator@0x7ff540021120 [IntegerLiteral@d2.swift:7:8]
  $T3 fully_bound subtype_of_existential involves_type_vars bindings={} @ locator@0x7ff540021208 [Call@d2.swift:7:3 -> function result]
  $T4 [lvalue allowed] subtype_of_existential involves_type_vars bindings={} @ locator@0x7ff5400212a0 [Call@d2.swift:7:3 -> apply function]
  $T5 fully_bound subtype_of_existential involves_type_vars bindings={} @ locator@0x7ff5400212a0 [Call@d2.swift:7:3 -> apply function]

Active Constraints:

Inactive Constraints:
  $T1 literal conforms to ExpressibleByIntegerLiteral [[locator@0x7ff540021080 [IntegerLiteral@d2.swift:7:5]]];
  $T2 literal conforms to ExpressibleByIntegerLiteral [[locator@0x7ff540021120 [IntegerLiteral@d2.swift:7:8]]];
  disjunction [[locator@0x7ff5400212a0 [Call@d2.swift:7:3 -> apply function]]]:$T4 bound to decl d2.(file).Callable.dynamicallyCall(withArguments:)@d2.swift:3:8 : (Callable) -> ([Int]) -> () at d2.swift:3:8 [[locator@0x7ff5400212a0 [Call@d2.swift:7:3 -> apply function]]]; or $T4 bound to decl d2.(file).Callable.dynamicallyCall(withArguments:)@d2.swift:4:8 : (Callable) -> ([Float]) -> () at d2.swift:4:8 [[locator@0x7ff5400212a0 [Call@d2.swift:7:3 -> apply function]]];
  $T5 conforms to ExpressibleByArrayLiteral [[locator@0x7ff5400212a0 [Call@d2.swift:7:3 -> apply function]]];
  $T5.ArrayLiteralElement can default to Any [[locator@0x7ff5400212a0 [Call@d2.swift:7:3 -> apply function]]];
  $T1 arg conv $T5.ArrayLiteralElement [[locator@0x7ff540021548 [Call@d2.swift:7:3 -> apply function -> tuple element #0]]];
  $T2 arg conv $T5.ArrayLiteralElement [[locator@0x7ff5400215d8 [Call@d2.swift:7:3 -> apply function -> tuple element #1]]];
  ($T5) -> $T3 applicable fn $T4 [[locator@0x7ff5400212a0 [Call@d2.swift:7:3 -> apply function]]];
```

These constraints seem sensible to me: `$T4` is bound to each method in the disjunction, and various constraints use it: `($T5) -> $T3 applicable fn $T4`, `$T5 conforms to ExpressibleByArrayLiteral`, etc.

Here are the runblocks I ran into:
1. One problem is that dynamic calls are not a pure syntactic sugar. Original call expressions have a tuple of multiple arguments (e.g. `x(1, 2)`), but the `dynamicallyCall` methods expect a single argument of array type (e.g. `x.dynamicallyCall(withArguments: [1, 2])`.

Thus, when the solver iterates over the options in the disjunction and tries to simplify the `($T5) -> $T3 applicable fn $T4` constraint, there is a crash in `matchCallArguments` because `x(1, 2)` does not fit the expected form `x.dynamicallyCall(WithArguments: [1, 2])`. The call site has two empty labels but the method has one label:
```
Assertion failed: (params.size() == labels.size()), function relabelParams, file /Users/dan/swift-build/swift/lib/AST/ASTContext.cpp, line 3623.
Stack dump:
...
swift::AnyFunctionType::relabelParams(llvm::MutableArrayRef<swift::AnyFunctionType::Param>, llvm::ArrayRef<swift::Identifier>) + 119
swift::constraints::matchCallArguments(swift::constraints::ConstraintSystem&, bool, llvm::ArrayRef<swift::AnyFunctionType::Param>, llvm::ArrayRef<swift::AnyFunctionType::Param>, swift::constraints::ConstraintLocatorBuilder) + 1616
swift::constraints::ConstraintSystem::simplifyApplicableFnConstraint(swift::Type, swift::Type, swift::OptionSet<swift::constraints::ConstraintSystem::TypeMatchFlags, unsigned int>, swift::constraints::ConstraintLocatorBuilder) + 1574
swift::constraints::ConstraintSystem::simplifyConstraint(swift::constraints::Constraint const&) + 1150
swift::constraints::ConstraintSystem::simplify(bool) + 210
swift::constraints::ConstraintSystem::solveRec(llvm::SmallVectorImpl<swift::constraints::Solution>&) + 83
swift::constraints::DisjunctionChoice::solve(llvm::SmallVectorImpl<swift::constraints::Solution>&) + 113
swift::constraints::ConstraintSystem::solveForDisjunctionChoices(swift::constraints::Disjunction&, llvm:)
```

I'm not sure how to work around this. One bad idea is to add a special case to `matchCallArguments` for resolving dynamic calls, but even this is difficult because when the solver iterates over the disjunction and tries to simplify the `($T5) -> $T3 applicable fn $T4` constraint, it's not known whether the call is a dynamic call or not.

2. Also, I'm not sure how to use disjunctions to handle tricker cases like this:
```swift
@dynamicCallable
struct S {
  func dynamicallyCall(withArguments: [String]) -> Int {
    fatalError()
  }
  func dynamicallyCall(withKeywordArguments: [String : String]) -> Float {
    fatalError()
  }
}

let s = S()
let _: Float = s("hi")
```

From what I understand, disjunction are useful for representing overloaded methods so that the solver can "iterate" over them and attempt to solve other constraints. However, that approach is insufficient by itself here because there are two orthogonal sets of constraints to test: constraints involving the `withArguments` method (e.g. `$T5 conforms to ExpressibleByArrayLiteral`) and ones involving the `withKeywordArguments` method (e.g. `$T5 conforms to ExpressibleByDictionaryLiteral`). I'm not sure how to reconcile the two: generating both sets of constraints at once is sure to fail.

---

Do you have suggestions for working around these roadblocks (1. that dynamic calls aren't an exact sugar, causing straightforward disjunction + `ApplicableFunction` constraints to fail, and 2. how to reconcile `withArguments` and `withKeywordArguments` constraints)?

Any ideas would be much appreciated!

rjmccall commented:
SE-0216 is quite clear about the ambiguity resolution rule:

> We propose that the type checker resolve this ambiguity towards the tightest match based on syntactic form of the expression. If a type implements both the `withArguments:` and `withKeywordArguments:` methods, the compiler will use the `withArguments:` method for call sites that have no keyword arguments and the `withKeywordArguments:` method for call sites that have at least one keyword argument.

The rule doesn't take into account anything except the presence of keyword arguments in the call site.  I'm concerned that this PR has gone off into the weeds trying to implement a much more complicated rule that isn't what was proposed, doesn't seem to have a strong motivation, and substantially complicates the user model of this feature.

lattner commented:
This is a great observation @rjmccall, your reading of the proposal is exactly right and (in retrospect) I remember the discussions that led to this - I'm sorry I page faulted on that and didn't remember this sooner.

I posted a note on the "amendment proposal" thread here:
https://forums.swift.org/t/amend-se-0216-dynamic-callable-reduce-overloads/16639/28

dan-zheng commented:
Thanks everyone for the help and insight!
I've revamped (and squashed) the implementation of `@dynamicCallable` in a fresh PR: https://github.com/apple/swift/pull/20305
Let us continue the discussion there.

dan-zheng commented:
FYI: [the new implementation PR was just merged.](https://github.com/apple/swift/pull/20305)
If you have feedback/suggestions, I'm happy to address them in master.
